### PR TITLE
AEA-1610 fix: add check consistency of package versions (#2459)

### LINKED ISSUE #2459: MAM  does not check consistency of package versions
**Is your feature request related to a problem? Please describe.**
Currently, we don't check for package version incompatibilities in MAM.

MAM should track all versions of packages and raise an exception when incompatible versions of any packages are added. Otherwise, this can lead to problems at runtime (with whatever version being loaded last being in memory).

**Describe the solution you'd like**
Provide a patch for the above.

Reproduce:
```
import os
from pathlib import Path
from aea.manager import MultiAgentManager

WORKING_DIR = "mam"

manager = MultiAgentManager(WORKING_DIR)
manager.start_manager()
from aea.configurations.base import PublicId
weather_station_id = PublicId.from_str("fetchai/weather_station:0.27.0")
manager.add_project(weather_station_id)
weather_client_id = PublicId.from_str("fetchai/weather_client:0.27.0")
manager.add_project(weather_client_id)
```
The second agent should not be addable, as it adds inconsistent versions (`"fetchai/weather_client:0.28.0"` is the compatible one)

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

Add consistency of package versions in MAM.

MAM should track all versions of packages and raise an exception when incompatible versions of any packages are added. Otherwise, this can lead to problems at runtime (with whatever version being loaded last being in memory).

This PR addresses the above issue.

## Fixes

Fix #2459 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [ ] Lint and unit tests pass locally with my changes and CI passes too
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

n/a

### CODE REVIEW

File: aea/configurations/data_types.py [hunk 1]
```diff
@@ -202,6 +202,9 @@ def __str__(self) -> str:
         return str(self.value)
 
 
+PackageIdPrefix = Tuple[ComponentType, str, str]
```
Comment (marcofavorito) [inline aea/configurations/data_types.py:205]: helper type, more readable, and does not affect much code

File: aea/manager/manager.py [hunk 2]
```diff
@@ -48,6 +50,38 @@ def __init__(self, msg: str, source_exception: Exception):
...
+        Initialize the exception.
+
+        :param agent_project_id: the agent project id whose addition has failed.
+        :param conflicting_packages: the conflicting packages.
+        """
+        self.agent_project_id = agent_project_id
+        self.conflicting_packages = conflicting_packages
+        super().__init__(self._build_error_message())
+
+    def _build_error_message(self) -> str:
```
Comment (marcofavorito) [inline aea/manager/manager.py:71]: for an example of the generated message, see below:
https://github.com/fetchai/agents-aea/pull/2550/files#diff-5a86c9151c53eb9b7fde346e231dc88a7b18f51ff369081b9116ef1c02423bddR604

File: tests/test_manager/test_manager.py [hunk 3]
```diff
@@ -585,6 +587,75 @@ class TestMultiAgentManagerThreadedModeWithPassword(BaseTestMultiAgentManager):
...
+        """
+        self.manager.start_manager()
+        weather_station_id = PublicId.from_str("fetchai/weather_station:0.27.0")
+        self.manager.add_project(weather_station_id)
+        weather_client_id = PublicId.from_str("fetchai/weather_client:0.27.0")
+        with pytest.raises(
+            ProjectPackageConsistencyCheckError,
+            match=re.escape(self.EXPECTED_ERROR_MESSAGE),
+        ):
+            self.manager.add_project(weather_client_id)
```
Comment (marcofavorito) [inline tests/test_manager/test_manager.py:647]: this is the code to reproduce issue https://github.com/fetchai/agents-aea/issues/2459

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Review by solarw — APPROVED** [review]
LGTM

File: aea/aea_builder.py [hunk 4]
```diff
@@ -116,9 +117,7 @@ def __init__(self) -> None:
         self._all_dependencies_by_type = (
             {}
         )  # type: Dict[ComponentType, Dict[ComponentId, ComponentConfiguration]]
-        self._prefix_to_components = (
-            {}
-        )  # type: Dict[Tuple[ComponentType, str, str], Set[ComponentId]]
+        self._prefix_to_components = {}  # type: Dict[PackageIdPrefix, Set[ComponentId]]
```
[the remark this answers is not in the document]
Comment (marcofavorito) [reply aea/aea_builder.py:120]: yes

**Pull request merged**